src/Requester/Decorator/BodyConverterDecorator.php

```diff
@@ -39,28 +45,46 @@ public function decorate(Request $request): Request
         $headers = new HeaderBag($request->getHeaders());
 
         if ($body !== null && ! is_string($body)) {
-            $doHandle = function () use ($body, $headers) {
+            $generator = function (?int $length = null) use ($body, $headers): Generator {
                 $body = $this->prepare($body);
 
                 if (is_iterable($body)) {
                     $body = self::encodeIterable($body, $headers);
                 }
 
-                return $body;
-            };
+                if ($length === null) {
+                    yield $body;
+
+                    return;
+                }
 
-            if (is_callable($body)) {
-                if (! $body instanceof Closure) {
-                    $body = Closure::fromCallable($body);
+                if (is_resource($body)) {
+                    while (! feof($body)) {
+                        yield fread($body, $length);
+                    }
+                } else {
+                    assert(is_string($body));
+
+                    $strlen = strlen($body);
+                    for ($i = 0; $i < $strlen; $i += $length) {
+                        yield substr($body, $i, $length);
+                    }
                 }
 
-                $refl = new ReflectionFunction($body);
-                $returnType = $refl->getReturnType();
-                // @phpstan-ignore-next-line
-                if ($returnType === null || $returnType->getName() !== 'string') {
-                    $body = $doHandle;
+                yield '';
+            };
+
+            $doHandle = static function (?int $length = null) use (&$generator) {
```

Comment:
`: string` as return type

Reply:
Not really always a string: if `$length` is null it could be a resource.

Anyway, the coding standard has a rule for that and should raise an error if a type hinting is missing. No need to check them manually anymore 🙂

Reply:
Ok, I'll resolve this one then